### ISSUE TITLE
feature request: token authentification

### ISSUE THREAD

**Issue by keiser1080** [zatosource/zato#546]
hi,

it will be great to have a token based authentification for the rest service.
It's more secure, and can be easy to implement.

Consurmer request a token using an authentification rest endpoint.
Then he can query a rest service using this token until the token expire.

Maybe the autentification endpoint can rely the other auth method for ex: basic auth, rbac ect...
And store the token to redis with a time to live.
And each service protected by token method can check if the token exist on redis.

**Comment by dsuch** [comment]
I get the idea and it sounds doable but I'm not sure what advantage it has over simple HTTP Basic Auth? Basic Auth is a username/password combination which is actually just a pair of tokens.

I understand that here the new thing is TTL stored for a token but can you give me some practical real-world scenarios where it would be useful?

What I mean is that if I understand it correctly, the workflow from the client's perspective would be:

* Obtain a token
* Use the token to call APIs
* If the token expired, request it again
* Continue to call APIs

In other words, clients need to constantly check if their calls are not rejected and if they are, they need to obtain the token again.

If that is the case, the Basic Auth is the same thing except that the token never expires so client's job is easier and I just don't know why it should expire at all.

Now, I'm not saying this is not a good idea - I'm sure it is! It's just that I've never had a need for it personally so I'm not sure when it would be used?

Thanks a lot @keiser1080!

**Comment by keiser1080** [comment]
Thanks for your answer
It's corret the flow is:
-     Obtain a token
-     Use the token to call APIs
-     If the token expired, request it again
-     Continue to call APIs

Rest is a stateless protocol.
Each request must containt all the datas.
So each time you reqyest a service you have to send your username & password.

It could be a security problem. If a request is intercepted  (ex:man on the middle)
Each request is one more chance to intercept the username password.

By using a token the username & password are less often send.
And if the token is intercepted, it can be used only until the next expire.

I am not an expert  :)

**Comment by dsuch** [comment]
Ok, I see - in that case there are two options:

* Use TLS and simply encrypt everything
* Wait for the next release which will have JSON Web Tokens (JWT) https://jwt.io/introduction/ which sounds is the thing you are thinking of

Actually TLS is always a good idea so nothing will prevent you from using JWT over TLS.

**Comment by dsuch** [comment]
Hi @keiser1080 - thank you for the ticket, I think JWT can handle your use-case so I will close the ticket.

**Comment by sebbean** [comment]
Would love token auth a la o auth! Is it possible even custom?
